### PR TITLE
core: local usage history / active documents 

### PR DESCRIPTION
**Pull request by ad-tra**
context: core
fixes: #1449

### CODE REVIEW

File: libs/core/src/lib.rs [hunk 1]
```diff
@@ -372,6 +372,14 @@ impl<Client: Requester> CoreLib<Client> {
         })
     }
 
+    #[instrument(level = "debug", skip(self), err(Debug))]
+    pub fn get_most_active_documents(&self) -> Result<Vec<(Uuid, usize)>, Error<GetUsageError>> {
```
Comment (ad-tra) [inline libs/core/src/lib.rs:376]: Hey, @SmailBarkouch, I exposed an endpoint in core `get_most_active_documents ` that you can use in android to improve the "Recent Documents" view. This endpoint will recommend more relevant files for the user.

The current approach for recommendation is to build a tree of all the files the user has access to, then sort them based on the newest. 
This new approach measures the count of read/write operations per document, then returns the most active documents (the ones with the most reads/writes).

This alternative approach will thus give more engaging file suggestions. How does this sound to you? Any editions or modifications you recommend?

File: libs/core/src/repo/schema_v3.rs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+hmdb::schema! {
+    CoreV3 {
+        account: <OneKey, Account>,
+        last_synced: <OneKey, i64>,
+        root: <OneKey, Uuid>,
+        local_metadata: <Uuid, SignedFile>,
+        base_metadata: <Uuid, SignedFile>,
+        read_activity: <Uuid, usize>,
+        write_activity: <Uuid, usize>,
```
Comment (Parth) [inline libs/core/src/repo/schema_v3.rs:27]: thinking about whether this should be 1 or 2 tables, also maybe pop these guys at the bottom

Comment (Parth) [reply libs/core/src/repo/schema_v3.rs:27]: I think 2 tables is fine

File: libs/core/src/repo/schema_v3.rs [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+                }
+                for (k, v) in source.username_by_public_key.get_all()? {
+                    tx.username_by_public_key.insert(k, v);
+                }
+                Ok::<_, hmdb::errors::Error>(())
+            })
+            .expect("failed to migrate local database from v2 to v3")?;
+        }
+
+        // migrate documents from id+source structure to id+hmac structure
```
Comment (Parth) [inline libs/core/src/repo/schema_v3.rs:68]: still a wip I assume because you didn't actually solicit a review, but v2 -> v3 will likely not need any of this stuff

**Comment by Parth** [comment]
overall looking good so far

File: libs/core/src/service/document_service.rs [hunk 4]
```diff
@@ -20,6 +20,9 @@ impl<Client: Requester> RequestContext<'_, '_, Client> {
 
         let doc = tree.read_document(self.config, &id, account)?;
 
+        let read_count = *self.tx.write_activity.get(&id).unwrap_or(&1);
```
Comment (SmailBarkouch) [inline libs/core/src/service/document_service.rs:23]: Did you mean to access `tx.read_activity` here?

File: libs/core/src/service/activity_service.rs [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+use crate::{CoreResult, RequestContext};
+use itertools::Itertools;
+use uuid::Uuid;
+
+impl<Client: Requester> RequestContext<'_, '_, Client> {
+    pub fn most_active_documents(&mut self) -> CoreResult<Vec<(Uuid, usize)>> {
+        let mut result: HashMap<Uuid, usize> = HashMap::new();
+
+        for (k, v) in self.tx.read_activity.get_all() {
+            result.insert(*k, if result.contains_key(k) { *v } else { *v + result[k] });
```
Comment (SmailBarkouch) [inline libs/core/src/service/activity_service.rs:13]: If `result` does contain a key already, wouldn't you add what it has? (in other words, I think you should take the inverse of your `if` statement)

File: libs/core/src/service/activity_service.rs [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+        }
+        for (k, v) in self.tx.write_activity.get_all() {
+            result.insert(*k, if result.contains_key(k) { *v } else { *v + result[k] });
```
Comment (SmailBarkouch) [inline libs/core/src/service/activity_service.rs:16]: Similar concern to the comment above.

File: libs/core/src/lib.rs (hunk 1 above)
Comment (SmailBarkouch) [reply libs/core/src/lib.rs:376]: I think this is a good approach, but I would somewhat argue there should be a balance between ranking what is accessed most, and what was most recently used. I feel this approach kinda makes "Recent Documents" into "Most Accessed Documents", as the most recently used file may not be accessed nearly as much (like a new file).

How this balance should be achieved is arguable. One approach I am thinking of is first ordering documents into "buckets" of time ranges (within 5 hour, 5 hour to 1 day, 1 day to 3 days, and so on), and within those ranges you do these read and write comparisons. To be honest this does complicate it a lot so definitely some other approach may be preferred.

File: libs/core/src/service/activity_service.rs (hunk 5 above)
Comment (SmailBarkouch) [reply libs/core/src/service/activity_service.rs:13]: To get rid of the if statement, you can use this: `result.get(k).unwrap_or_else(0) + *v`.

Comment (ad-tra) [reply libs/core/src/service/activity_service.rs:13]: ty for the oneliner!

**Review by Parth — CHANGES_REQUESTED** [review]

File: libs/core/src/repo/schema_v3.rs [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        read_activity: <Uuid, usize>,
+        write_activity: <Uuid, usize>
+    }
+}
+
+impl CoreV3 {
+    pub fn init_with_migration(writeable_path: &str) -> Result<CoreV3, CoreError> {
+        let db = CoreV3::init(writeable_path)?;
+
+        // migrate metadata from v1
```
Comment (Parth) [inline libs/core/src/repo/schema_v3.rs:35]:
```suggestion
        // migrate metadata from v2
```

**Review by Parth — CHANGES_REQUESTED** [review]
Additionally as we talked about, let's use a richer data model right off the bat, once you have that you can answer the 4 questions we talked about:

{most frequent, most recently} x {read, wrote}

and we can play around with some suggestive algorithm like @SmailBarkouch was thinking about.

File: libs/core/src/service/activity_service.rs [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+        Ok(scores.into_iter().map(|f| f.0).collect_vec())
+    }
+
+    fn normalize(scores: &mut [(Uuid, DocActivityMetrics)]) {
+        let mut docs_avg_read_timestamps = scores
+            .iter_mut()
+            .map(|f| f.1.avg_read_timestamp)
+            .collect_vec();
+        docs_avg_read_timestamps.sort_by(|a, b| a.raw.cmp(&b.raw));
```
Comment (Parth) [inline libs/core/src/service/activity_service.rs:38]: I wouldn't sort these to just get the min & max. That's `nlogn` vs `n`. I expect these to be reasonably large collections.

File: libs/core/src/service/document_service.rs [hunk 9]
```diff
@@ -20,6 +23,23 @@ impl<Client: Requester> RequestContext<'_, '_, Client> {
 
         let doc = tree.read_document(self.config, &id, account)?;
 
+        let mut doc_events = self.tx.docs_events.get(&id).unwrap_or(&Vec::new()).clone();
+        doc_events.sort();
+        let latest_event = doc_events
+            .iter()
+            .filter(|e| matches!(e, DocEvents::Read(_)))
+            .last();
```
Comment (Parth) [inline libs/core/src/service/document_service.rs:31]: I would use this [`last()`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.last) instead, O(n) vs O(1)

Comment (Parth) [reply libs/core/src/service/document_service.rs:31]: You don't have to filter for the last read, you can just check if the last is a read, same goes for write.

**Review by Parth — CHANGES_REQUESTED** [review]

File: libs/core/src/lib.rs [hunk 10]
```diff
@@ -293,6 +293,10 @@ impl<Client: Requester> CoreLib<Client> {
         })
     }
 
+    #[instrument(level = "debug", skip(self), err(Debug))]
+    pub fn get_most_active_documents(&self) -> Result<Vec<Uuid>, Error<GetUsageError>> {
+        Ok(self.in_tx(|s| s.suggested_docs())?)
```
Comment (Parth) [inline libs/core/src/lib.rs:298]: should call these the same thing

File: libs/core/src/repo/mod.rs [hunk 11]
```diff
@@ -24,6 +25,7 @@ pub struct CoreV3 {
     pub local_metadata: LookupTable<Uuid, SignedFile>,
     pub base_metadata: LookupTable<Uuid, SignedFile>,
     pub pub_key_lookup: LookupTable<Owner, String>,
+    pub doc_events: LookupTable<Uuid, Vec<DocEvents>>,
```
Comment (Parth) [inline libs/core/src/repo/mod.rs:28]: You should add a `LookupList` to `db_rs`

File: libs/core/src/repo/schema_v3.rs [hunk 12]
```diff
@@ -0,0 +1,69 @@
+use crate::CoreError;
```
Comment (Parth) [inline libs/core/src/repo/schema_v3.rs:1]: this whole file can / should go away

File: libs/core/src/service/activity_service.rs [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+        Ok(scores.into_iter().map(|f| f.0).collect_vec())
+    }
+
+    fn normalize(scores: &mut [(Uuid, DocActivityMetrics)]) {
+        let docs_avg_read_timestamps = StatisticValueRange {
+            max: scores
+                .iter_mut()
+                .map(|f| f.1.avg_read_timestamp)
+                .max()
+                .unwrap(),
```
Comment (Parth) [inline libs/core/src/service/activity_service.rs:36]: no unwraps, I believe if a doc has no activity this will crash

Comment (Parth) [reply libs/core/src/service/activity_service.rs:36]: you can `unwrap_or` or map it to an error

File: libs/core/src/service/activity_service.rs [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+            min: scores
+                .iter_mut()
+                .map(|f| f.1.avg_read_timestamp)
+                .min()
+                .unwrap(),
+        };
+
+        let docs_avg_write_timestamps = StatisticValueRange {
+            max: scores
+                .iter_mut()
```
Comment (Parth) [inline libs/core/src/service/activity_service.rs:46]: why are these `iter_mut`s?

File: libs/core/src/service/document_service.rs [hunk 15]
```diff
@@ -41,6 +65,27 @@ impl<Client: Requester> CoreState<Client> {
         let hmac = tree.find(&id)?.document_hmac();
         document_repo::insert(&self.config, &id, hmac, &encrypted_document)?;
 
+        let mut doc_events = self
+            .db
+            .doc_events
+            .data()
+            .get(&id)
+            .unwrap_or(&Vec::new())
+            .clone();
```
Comment (Parth) [inline libs/core/src/service/document_service.rs:74]: could be an expensive operation

File: libs/core/src/service/document_service.rs [hunk 16]
```diff
@@ -41,6 +65,27 @@ impl<Client: Requester> CoreState<Client> {
...
+
+        doc_events.sort_by(|a, b| b.cmp(a)); //sort in descending order
```
Comment (Parth) [inline libs/core/src/service/document_service.rs:76]: could be an expensive operation

Comment (Parth) [reply libs/core/src/service/document_service.rs:76]: can't you just keep the lists sorted by time?

File: libs/core/src/service/document_service.rs [hunk 17]
```diff
@@ -41,6 +65,27 @@ impl<Client: Requester> CoreState<Client> {
...
+        let latest_event = doc_events.iter().find(|e| matches!(e, DocEvents::Read(_)));
```
Comment (Parth) [inline libs/core/src/service/document_service.rs:77]: i think you just need to peak at the last element and determine whether enough time has elapsed before writing

File: libs/core/src/service/document_service.rs [hunk 18]
```diff
@@ -41,6 +65,27 @@ impl<Client: Requester> CoreState<Client> {
...
+
+        let is_capped = match latest_event {
+            Some(event) => Utc::now().timestamp() - event.timestamp() > RATE_LIMIT_MILLIS,
+            None => true,
+        };
+
+        if !is_capped {
+            doc_events.push(document_repo::DocEvents::Read(Utc::now().timestamp()));
+            self.db.doc_events.insert(id, doc_events)?;
```
Comment (Parth) [inline libs/core/src/service/document_service.rs:86]: This little interaction should entirely inform the db-rs `LookupList`

File: libs/core/src/service/document_service.rs [hunk 19]
```diff
@@ -19,6 +21,28 @@ impl<Client: Requester> CoreState<Client> {
             .ok_or(CoreError::AccountNonexistent)?;
 
         let doc = tree.read_document(&self.config, &id, account)?;
+        self.db.doc_events.data();
```
Comment (Parth) [inline libs/core/src/service/document_service.rs:24]: same comments as write

File: libs/core/src/service/document_service.rs (hunk 17 above)
Comment (Parth) [reply libs/core/src/service/document_service.rs:77]: there is also a copy paste related mistake here, avoid copy pasting things

**Review by Parth — CHANGES_REQUESTED** [review]

**Review by ad-tra — COMMENTED** [review]
- [x] [data model change](https://discord.com/channels/@me/1029027220179669004/1075120834961756180) use a list instead of linked list 
- [x] uniform use of iter 
- [ ] `///`
- [x] last access instead of averages

File: libs/core/libs/shared/src/document_repo.rs [hunk 20]
```diff
@@ -1,15 +1,155 @@
...
+pub struct DocActivityMetrics {
+    pub id: Uuid,
+    pub latest_read_timestamp: StatisticValue,
+    pub latest_write_timestamp: StatisticValue,
+    pub read_count: StatisticValue,
+    pub write_count: StatisticValue,
+}
+impl DocActivityMetrics {
+    pub fn rank(&self) -> i64 {
+        (self.latest_read_timestamp.normalized + self.latest_write_timestamp.normalized) as i64 * 70
```
Comment (ad-tra) [inline libs/core/libs/shared/src/document_repo.rs:85]: @Parth, What do you think about exposing the weights (70%, 30%) in the core endpoint as parameters to give flexibility to the caller? That way, this endpoint can serve as both suggested documents and most recent (if you pass the timestamp weight to 0), most accessed(if you pass the write/read count as 0 )

File: server/server/src/file_service.rs
Comment (ad-tra) [inline server/server/src/file_service.rs:1]: The change in this file is unrelated to the pr main goal. Clippy was complaining, so I just applied its suggestions.

File: server/server/src/utils.rs
Comment (ad-tra) [inline server/server/src/utils.rs:1]: same as `file_service.rs`

File: libs/core/libs/shared/src/document_repo.rs (hunk 20 above)
Comment (Parth) [reply libs/core/libs/shared/src/document_repo.rs:86]: I think we should expose these weights when clients want to customize them / expose them to users, for now it's fine

File: libs/core/libs/shared/src/document_repo.rs [hunk 21]
```diff
@@ -1,15 +1,156 @@
...
+    pub min: StatisticValue,
+}
+impl StatisticValue {
+    pub fn normalize(&mut self, range: StatisticValueRange) {
+        self.normalized = (self.raw - range.min.raw) as f64 / (range.max.raw - range.min.raw) as f64
+    }
+}
+/// DocActivityMetrics stores key document activity features, which are used to recommend relevant documents to the user.
+/// Here's a walkthrough of the recommendation procedure: collect 1k most recent document events (write/read), use that activity to construct a DocActivtyMetrics struct for each document. Min-max normalizes the activity features, then rank the documents.
+/// latest_read_timestamp: the latest epoch timestamp that the user read a document
```
Comment (Parth) [inline libs/core/libs/shared/src/document_repo.rs:72]: you could just add these as field level comments, but not a huge deal

File: libs/core/libs/shared/src/document_repo.rs [hunk 22]
```diff
@@ -1,15 +1,156 @@
...
+}
+
+#[derive(Default, Copy, Clone, PartialEq)]
+pub struct StatisticValue {
+    pub raw: i64,
+    pub normalized: f64,
+}
+impl StatisticValue {
+    fn from_raw(raw: i64) -> StatisticValue {
+        StatisticValue { raw, normalized: f64::default() }
```
Comment (Parth) [inline libs/core/libs/shared/src/document_repo.rs:43]: I would personally either
1. not have this exist until you can calculate the normalized value
2. make `normalized: Option<f64>` to represent it's un-calculated state

usually I try to do 1

File: libs/core/libs/shared/src/document_repo.rs [hunk 23]
```diff
@@ -1,15 +1,156 @@
...
+    pub id: Uuid,
+    pub latest_read_timestamp: StatisticValue,
+    pub latest_write_timestamp: StatisticValue,
+    pub read_count: StatisticValue,
+    pub write_count: StatisticValue,
+}
+impl DocActivityMetrics {
+    pub fn rank(&self) -> i64 {
+        (self.latest_read_timestamp.normalized + self.latest_write_timestamp.normalized) as i64 * 70
+            + (self.read_count.normalized + self.write_count.normalized) as i64 * 30
```
Comment (Parth) [inline libs/core/libs/shared/src/document_repo.rs:87]: I would however extract these as constants

File: libs/core/libs/shared/src/document_repo.rs [hunk 24]
```diff
@@ -1,15 +1,156 @@
...
+#[derive(Default, Copy, Clone)]
+pub struct DocActivityMetrics {
+    pub id: Uuid,
+    pub latest_read_timestamp: StatisticValue,
+    pub latest_write_timestamp: StatisticValue,
+    pub read_count: StatisticValue,
+    pub write_count: StatisticValue,
+}
+impl DocActivityMetrics {
+    pub fn rank(&self) -> i64 {
```
Comment (Parth) [inline libs/core/libs/shared/src/document_repo.rs:85]: rank usually indicates a position in a list (think valedictorian). score is what I think you're going for here (think gpa).

File: libs/core/libs/shared/src/document_repo.rs [hunk 25]
```diff
@@ -1,15 +1,156 @@
...
+                None => {
+                    set.insert(event.id(), vec![event]);
+                }
+                Some(events) => {
+                    events.push(event);
+                }
+            }
+        }
+
+        for pair in set {
```
Comment (Parth) [inline libs/core/libs/shared/src/document_repo.rs:112]:
```suggestion
        for (key, value) in set {
```

Comment (Parth) [reply libs/core/libs/shared/src/document_repo.rs:112]: then you don't have to do the `.1` `.0` business

Comment (Parth) [reply libs/core/libs/shared/src/document_repo.rs:112]: can probably do one step better than `key`, `value`

File: libs/core/libs/shared/src/document_repo.rs [hunk 26]
```diff
@@ -1,15 +1,156 @@
...
+            let read_events: Vec<&&DocEvent> = pair
+                .1
+                .iter()
+                .filter(|e| matches!(e, DocEvent::Read(_, _)))
+                .collect();
+            let latest_read = read_events
+                .iter()
+                .max_by(|x, y| x.timestamp().cmp(&y.timestamp()));
```
Comment (Parth) [inline libs/core/libs/shared/src/document_repo.rs:120]: can't you just say it's the _last_ read?

File: libs/core/libs/shared/src/document_repo.rs [hunk 27]
```diff
@@ -1,15 +1,156 @@
...
+                    set.insert(event.id(), vec![event]);
+                }
+                Some(events) => {
+                    events.push(event);
+                }
+            }
+        }
+
+        for pair in set {
+            let read_events: Vec<&&DocEvent> = pair
```
Comment (Parth) [inline libs/core/libs/shared/src/document_repo.rs:113]: I don't think you actually need random access to the read events, you just need either the `.count()` or the `.last()` element. `collect()`'ing into a `vec` is an expensive allocation that it doesn't look like you benefit from here.

File: libs/core/src/service/activity_service.rs [hunk 28]
```diff
@@ -0,0 +1,71 @@
...
+            max: self.map(|f| f.latest_write_timestamp).max().unwrap(),
+            min: self.map(|f| f.latest_write_timestamp).min().unwrap(),
+        };
+        self.for_each(|f| {
+            f.read_count.normalize(read_count_range);
+            f.write_count.normalize(write_count_range);
+            f.latest_read_timestamp.normalize(latest_read_range);
+            f.latest_write_timestamp.normalize(latest_write_range);
+        });
+        vec![]
```
Comment (Parth) [inline libs/core/src/service/activity_service.rs:69]: looks like this is a prototyping mistake?

File: libs/core/src/service/activity_service.rs [hunk 29]
```diff
@@ -0,0 +1,71 @@
...
+    fn normalize(&mut self) -> Vec<DocActivityMetrics>;
+}
+impl<'a, T> Normalizer for T
+where
+    T: Iterator<Item = &'a mut DocActivityMetrics>,
+{
+    fn normalize(&mut self) -> Vec<DocActivityMetrics> {
+        let read_count_range = StatisticValueRange {
+            max: self.map(|f| f.read_count).max().unwrap(),
+            min: self.map(|f| f.read_count).min().unwrap(),
```
Comment (Parth) [inline libs/core/src/service/activity_service.rs:48]: no unwraps!

File: libs/core/src/service/activity_service.rs [hunk 30]
```diff
@@ -0,0 +1,71 @@
...
+use db_rs::DbError;
+use lockbook_shared::document_repo::DocActivityMetrics;
+use lockbook_shared::document_repo::DocEvent;
+use lockbook_shared::document_repo::StatisticValueRange;
+use lockbook_shared::document_repo::Stats;
+use uuid::Uuid;
+
+impl<Client: Requester> CoreState<Client> {
+    pub(crate) fn suggested_docs(&mut self) -> LbResult<Vec<Uuid>> {
+        let mut scores = self
```
Comment (Parth) [inline libs/core/src/service/activity_service.rs:13]: with the current implemtnation, I'm pretty sure `score == vec![]` no matter what the current state of things is. And in some situations you're going to panic. You need some tests in these parts to catch these mistakes.

File: libs/core/src/service/activity_service.rs [hunk 31]
```diff
@@ -0,0 +1,71 @@
...
+            .db
+            .doc_events
+            .data()
+            .iter()
+            .get_activity_metrics()
+            .iter_mut()
+            .normalize();
+
+        scores.sort_by(|a, b| DocActivityMetrics::rank(a).cmp(&DocActivityMetrics::rank(b)));
```
Comment (Parth) [inline libs/core/src/service/activity_service.rs:22]: I think it makes sense to sort_by this by default

**Review by Parth — CHANGES_REQUESTED** [review]

File: libs/core/libs/shared/src/document_repo.rs (hunk 27 above)
Comment (ad-tra) [reply libs/core/libs/shared/src/document_repo.rs:113]: it's either collecting into a vec or cloning the iter.

**Review by Parth — APPROVED** [review]
